src/typechecker/FStar.TypeChecker.Cfg.fst

```diff
@@ -1063,17 +1090,104 @@ let built_in_primitive_steps : prim_step_set =
                 Some body
               | _ -> None)
             (fun body -> body)
-            (fun _t body -> Some body)
+            (fun _us _t body -> Some body)
+    in
+    let array_ops =
+      let of_list_op =
+        let emb_typ t = ET_app(PC.immutable_array_t_lid |> Ident.string_of_lid, [t]) in
+        let un_lazy universes t l r =
+          S.mk_Tm_app
+            (S.mk_Tm_uinst (U.fvar_const PC.immutable_array_of_list_lid) universes)
+            [S.iarg t; S.as_arg l]
+            r
+        in
+        (  PC.immutable_array_of_list_lid, 2, 1,
+           mixed_binary_op
+              (fun (elt_t, _) -> Some elt_t)
+              (fun (l, q) ->
+                match arg_as_list FStar.Syntax.Embeddings.e_any (l, q) with
+                | Some lst -> Some (l, lst)
+                | _ -> None)
+              (fun r (universes, elt_t, (l, blob)) ->
+                S.mk (Tm_lazy { blob;
+                                lkind=Lazy_embedding (emb_typ EMB.(emb_typ_of e_any), Thunk.mk (fun _ -> un_lazy universes elt_t l r));
+                                ltyp=S.mk_Tm_app (S.mk_Tm_uinst (U.fvar_const PC.immutable_array_t_lid) universes) [S.as_arg elt_t] r;
+                                rng=r }) r)
+              (fun r universes elt_t (l, lst) ->
+                 let blob = FStar.ImmutableArray.Base.of_list #term lst in
+                 Some (universes, elt_t, (l, FStar.Compiler.Dyn.mkdyn blob))),
+           NBETerm.mixed_binary_op
+             (fun (elt_t, _) -> Some elt_t)
+             (fun (l, q) ->
+               match NBETerm.arg_as_list NBETerm.e_any (l, q) with
+               | None -> None
+               | Some lst -> Some (l, lst))
+             (fun (universes, elt_t, (l, blob)) ->
+               NBETerm.mk_t <|
+               NBETerm.Lazy (Inr (blob, emb_typ EMB.(emb_typ_of e_any)),
+                             Thunk.mk (fun _ ->
+                               NBETerm.mk_t <| NBETerm.FV (S.lid_as_fv PC.immutable_array_of_list_lid S.delta_constant None,
```

Comment:
The NBE version is similar, except the thunk represents the unreduced `of_list l` as an NBE.t